### ISSUE TITLE
Latex inside $..$ tags...

### ISSUE THREAD

**Issue by pasitolonen** [jgm/pandoc#1862]
We have a web application that outputs HTML code to further processing (pandoc) and we're including our own plain Latex code within $...$ tags using parameter
```
-f html+tex_math_dollars
```
However, we have realized that it is somehow a limited approach since we get into errors with simple commands like \newpage etc. I know this $...$ approach is intended for math, but is there a way to let it pass everything between dollar signs as pure Latex.

**Comment by ousia** [comment]
Maybe [Raw TeX](http://pandoc.org/README.html#raw-tex) helps here.

**Comment by pasitolonen** [comment]
Thanks. I'll give it a try. Hence, it says that "Inline LaTeX is ignored in output formats other than Markdown, LaTeX, and ConTeXt." and we're outputting pdf.

**Comment by pasitolonen** [comment]
This is a minimum working example. Consider the following code as file test.html
```
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>My Test File</title>
</head>
<body>
This is \underline{underlined} text.
</body>
</html>
```
And the command line...
```
pandoc test.html -f html+raw_tex -t latex -o test.pdf
```
Doesn't compile Latex underline command as underlined text rather than passes it through as regular text.

**Comment by jgm** [comment]
PDF output is via latex, so raw tex should work.

+++ pasitolonen [Jan 07 15 00:31 ]:

**Comment by jgm** [comment]
Oh, I see, your source is HTML, not markdown.  +raw_tex only works with Markdown.

Why do you want to include raw LaTeX in your HTML (other than math)?

+++ pasitolonen [Jan 07 15 03:58 ]:

**Comment by pasitolonen** [comment]
John,

Our web application stores/shows end user documents in HTML and we use Pandoc + templates to convert them via Latex to final PDFs. We can pass most of the raw Latex commands required for additional typesetting (other than we have in templates) using current $...$ tags. For example, we need conditional formatting in certain paragraphs and that has to be raw Latex like page breaks, indents, margins etc. We have managed to hack our way around most of the problems, but this can't be final solution.

Therefore, we are in need of this "raw Latex" feature either by extending the way $..$ tags work or some other manner.

**Comment by pasitolonen** [comment]
Continued. Since Haskell is a foreign language to me, I didn't want to dive into coding right away, but I guess it might not be impossible to extend **tex_math_dollars** so that you can have, for example, like **tex_math_dollars_raw** with what you just pass everything through.

**Comment by jgm** [comment]
I'd suggest using some HTML div and span tags to encode the page formatting information you need.  You can easily intercept these in a pandoc filter and convert them to the LaTeX you want.  So, e.g.

```
<div class="pagebreak"/>
```

See the documentation on pandoc filters.

+++ pasitolonen [Jan 07 15 10:11 ]:

**Comment by pasitolonen** [comment]
John, thank you for this idea. I'll start exploring it right away.

**Comment by pasitolonen** [comment]
I made a simple test. The source file, **test.html** is as follows:

````
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <title>My Test File</title>
</head>

<body>
<div class="rawlatex">
   \underline{No we can't!}
</div>
</body>
</html>
```
And the pretty near exact copy of theorem.py filter as **rawlatex.py** is as follows:
```
#!/usr/bin/env python

"""
Pandoc filter to convert divs with class="theorem" to LaTeX
theorem environments in LaTeX output, and to numbered theorems
in HTML output.
"""

from pandocfilters import toJSONFilter, RawBlock, Div, Str


def latex(x):
    return RawBlock('latex', x)


def html(x):
    return RawBlock('html', x)


def rawfilter(key, value, format, meta):
    if key == 'Div':
        [[ident, classes, kvs], contents] = value
        if "rawlatex" in classes:
            if format == "latex":
                return(contents)
            else:
                return []

if __name__ == "__main__":
    toJSONFilter(rawfilter)
```
My command line is
```
pandoc -s --filter ./rawlatex.py test.html -f html -t latex -o test.tex
```
It compiles ok, but still - in my .tex file I have 
```
\textbackslash{}underline\{No we can't!\}
```
I must be missing something...

**Comment by jgm** [comment]
Where you have `return(contents)`, try:

    return latex(contents)

By the way, my original idea was different -- that you would
not put raw LaTeX in the HTML at all, but rather
*substitute* for the HTML tag in the filter.

So, if you want underlining, for example, you'd
use

    <span class="underline">foo</span>

in your HTML.  Then your filter would do something like


    if key == 'Span':
      [[ident, classes, kvs], contents] = value
      if "underline" in classes and format == "latex":
        return ([RawInline('latex', '\\underline{')] +
                   contents + [RawInline('latex', '}')]

+++ pasitolonen [Jan 07 15 13:19 ]:

**Comment by pasitolonen** [comment]
I tried that first, but it gives
```
pandoc: when expecting a String, encountered Array instead
```

**Comment by jgm** [comment]
Post your whole script (the one that gives this error), and someone here
can help you debug it.

+++ pasitolonen [Jan 07 15 13:38 ]:

**Comment by pasitolonen** [comment]
John,

I truly appreciate your help, thank you! As you suggested, the principle that all *raw* Latex commands are constructed from respective HTML-spans works well. 

Source file **test.html**
```
<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <title>My Test File</title>
</head>

<body>
<span class="underline">
   Yes we can!
</span>
   No we can't
</body>
</html>
```
As an example, contents of <span class "underline> are converted to Latex command \underline{contents of span}. Pandoc filter **rawhtml.py**. 
```
#!/usr/bin/env python

from pandocfilters import toJSONFilter, RawInline

def rawfilter(key, value, format, meta):
    if key == 'Span':
        [[ident, classes, kvs], contents] = value
        if "underline" in classes and format == "latex":
            return ([RawInline('latex', '\\underline{')] + contents + [RawInline('latex', '}')])

if __name__ == "__main__":
    toJSONFilter(rawfilter)
```
The following command line constructs a PDF.
```
$ pandoc -s --filter ./rawlatex.py test.html -f html -t latex -o test.pdf
```
we'll get properly formatted PDF in which the text "Yes we can!" is underlined as supposed.

Bottom line. Pandoc is a fantastic tool!
